R/set-record-option.R

```diff
@@ -0,0 +1,110 @@
+
+#' Add or modify a record option
+#'
+#' Add a new record option or adjust the value/presence of an existing one.
+#'
+#' Note that only estimation, data, and table records are currently supported.
+#'
+#' Notes on `value`:
+#'  - We infer whether to use `option_value` or `option_flag` for *new* options based on the *class* of `value`:
+#'     - If you are trying to add an option with an assigned value (i.e. `MAXEVAL = 10`), `value` must be a character string or numeric value
+#'     - If you are trying to add a flag option (i.e. `NOABORT`), `value` should be logical (`T`/`F`).
+#'  - If `value` is set to `NULL`, the option will be removed (if found). `FALSE` is also acceptable for removing an existing option of class `nmrec_option_flag`.
+#'
+#' @param record An [nmrec_record] object.
+#' @param name Name of record to select. Any valid spelling of the option name
+#'   is allowed.
+#' @param value value to assign to option.
+#'
+#' @seealso `nmrec::option` selecting records of a given type.
+#'
+#'
+set_record_option <- function(record,
+                              name,
+                              value) {
+  stopifnot(inherits(record, "nmrec_record"))
+  stopifnot(length(name) == 1, is.character(name), nzchar(name))
+  stopifnot(record[["name"]] %in% c("estimation", "data", "table"))
+
+
+  name_map <- get(paste0(record[["name"]], "_option_names"))
+  name_resolved <- get0(tolower(name), name_map)
+  rec_opt_prev <- get_record_option(record, name)
+
+
+  if (is.null(name_resolved) && is.null(rec_opt_prev)) {
+    abort(
+      sprintf("%s is not a valid option for %s blocks", name, record[["name"]]),
+      nmrec_error()
+    )
+  }
+
+  # Create new record option
+  if (is.null(rec_opt_prev)) {
+    if (is.null(value)) {
+      abort(
+        sprintf("Option %s does not exist in this record. `NULL` is only valid for removing existing options.", name),
+        nmrec_error()
+      )
+    } else {
+      opt_type <- if (is.logical(value)) "flag" else "value"
+
+      if (opt_type == "value") {
+        stopifnot(is.numeric(value) || is.character(value))
+        opt <- option_value$new(name_resolved, name, value = value)
+      } else if (opt_type == "flag") {
+        stopifnot(is.logical(value))
+        opt <- option_flag$new(name_resolved, name)
+      }
+    }
+
+    new_opt_lst <- list(elem_whitespace(" "), opt)
+
+    record$parse()
+    record$values <- append(record$values, new_opt_lst, after = find_record_end(record) - 1)
```

Comment:
I don't think `find_record_end` is necessary.  You can get the value with something like

```R
purrr::detect_index(
  record$values,
  function(x) !elem_is(x, "linebreak"),
  .dir = "backward"
)
```

Reply:
That doesnt quite accomplish what I was trying to achieve, as you can see by the `while` statement (maybe a better way to do that though). One of the example `EST` records ended with two new lines, and I didnt want an empty line to be between them. In those cases we need one newline back from the final one. I think we'd still want to do this even if we implemented this [suggestion](https://github.com/metrumresearchgroup/nmrec/pull/5#discussion_r1227142812) to ensure there are no empty line separators.

Reply:
Please show the specific case where that wouldn't work.

Reply:
What I meant by that is that it wouldn't return the index I was looking for, not that we couldnt use this method. Below is the scenario I was referring to.
```r
record$values

...

[[30]]
[1] "\n"
attr(,"class")
[1] "nmrec_linebreak" "nmrec_element"   "character"      

[[31]]
[1] "\n"
attr(,"class")
[1] "nmrec_linebreak" "nmrec_element"   "character"      

> purrr::detect_index(
+     record$values,
+     function(x) !elem_is(x, "linebreak"),
+     .dir = "backward"
+ )
[1] 29
> find_record_end(record)
[1] 30
```

In terms of its specific use within `set_record_option`, it seems fine to make the substitution. We would loose the ability to find that index (which I think is valuable for other purposes too, and adding 1 to the `purrr` method doesn't have the same guarantees it's the first new line), but maybe that's not worth including until a _real_ need arises.

Reply:
Sorry, you seem to have specific cases in mind when you're talking about "guarantees", but I'm not following.

Anyway, you seem to now agree that passing the result of the `purrr::detect_index` snippet I gave as `after` would give the same result as passing `find_record_end(record) - 1`, so yes, let's not add a custom function when a `detect_index` call will do.